Refactor caret painting logic

mwyrzykowski opened the pull request:
#### e314f573db05b5bdbed3b794661dc0144f32ba16
<pre>
Refactor caret painting logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=252148">https://bugs.webkit.org/show_bug.cgi?id=252148</a>
&lt;radar://104975415&gt;

Reviewed by Ryosuke Niwa and Aditya Keerthi.

Continue previous refactorings to the caret painting
logic to allow some additional flexibility.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::fillCaretRect):
Pass not just the presentation properties but the entire
animator to allow for additional flexibility.

(WebCore::repaintCaretRectForLocalRect):
(WebCore::repaintCaretForLocalRect):
Allow the animator to influence the repaint rect.

(WebCore::FrameSelection::invalidateCaretRect):
(WebCore::CaretBase::invalidateCaretRect):
(WebCore::FrameSelection::paintCaret):
(WebCore::CaretBase::paintCaret const):
(WebCore::DragCaretController::paintDragCaret const):
* Source/WebCore/editing/FrameSelection.h:
Pass the caret animator instead of just its properties.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::renderCaretInsideContentsClip):
(WebCore::RenderBlock::paint):
(WebCore::RenderBlock::paintObject):
Allow rendering the caret outside the element&apos;s clipping rectangle.

Canonical link: <a href="https://commits.webkit.org/260775@main">https://commits.webkit.org/260775@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1832b987e2206700c1d14fea1c6f8be6b0ede6b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109322 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18401 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118533 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113205 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9686 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101559 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115077 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14872 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98103 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43059 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96844 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84822 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11172 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31099 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11909 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50702 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7455 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13526 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -1120,6 +1129,11 @@ void RenderBlock::paint(PaintInfo& paintInfo, const LayoutPoint& paintOffset)
     if (pushedClip)
         popContentsClip(paintInfo, phase, adjustedPaintOffset);
 
+    if (paintInfo.phase == PaintPhase::Foreground && !renderCaretInsideContentsClip()) {
```
@smfr I want to render the caret outside of the RenderBlock / text field's clipping rectangle. Is there a better way to do this than ignoring the clipping rectangle? Should I paint the caret as part of a different paint phase?

pxlcoder commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -1120,6 +1129,11 @@ void RenderBlock::paint(PaintInfo& paintInfo, const LayoutPoint& paintOffset)
 
+    if (paintInfo.phase == PaintPhase::Foreground && !renderCaretInsideContentsClip()) {
+        paintCaret(paintInfo, adjustedPaintOffset, CursorCaret);
+        paintCaret(paintInfo, adjustedPaintOffset, DragCaret);
```
Can we factor these two lines into a separate method, since it's duplicated below?

mwyrzykowski replied:
Good idea, I'll make that change

mwyrzykowski committed:
Refactor caret painting logic
https://bugs.webkit.org/show_bug.cgi?id=252148
<radar://104975415>

Reviewed by Ryosuke Niwa and Aditya Keerthi.

Continue previous refactorings to the caret painting
logic to allow some additional flexibility.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::fillCaretRect):
Pass not just the presentation properties but the entire
animator to allow for additional flexibility.

(WebCore::repaintCaretRectForLocalRect):
(WebCore::repaintCaretForLocalRect):
Allow the animator to influence the repaint rect.

(WebCore::FrameSelection::invalidateCaretRect):
(WebCore::CaretBase::invalidateCaretRect):
(WebCore::FrameSelection::paintCaret):
(WebCore::CaretBase::paintCaret const):
(WebCore::DragCaretController::paintDragCaret const):
* Source/WebCore/editing/FrameSelection.h:
Pass the caret animator instead of just its properties.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::renderCaretInsideContentsClip):
(WebCore::RenderBlock::paint):
(WebCore::RenderBlock::paintObject):
Allow rendering the caret outside the element's clipping rectangle.

Canonical link: https://commits.webkit.org/260775@main
